### ISSUE TITLE
Connection to relay servers should resume TLS sessions on re-connect

### ISSUE THREAD

**Issue by flub** [n0-computer/iroh#3083]
The relay client in the MagicSock sometimes wants to reconnect to be sure the connection is good.  When doing so however it does not use the RLS session resumption, here the relevant rustls logs:

```
rustls::client::hs: 83: No cached session for IpAddress(V4(Ipv4Addr([127, 0, 0, 1])))
rustls::client::hs: 141: Not resuming any session
rustls::client::hs: 532: Sending ClientHello Message
...
```

This despite the server issuing tickets after the connection is established:

```
rustls::server::tls13: 1342: sending new ticket HandshakeMessagePayload 
```

It would be good if we could re-use this ticket and setup the subsequent connections faster.

**Comment by DXist** [comment]
0-rtt handshake is not a TLS session resumption - TLS1.3 dropped support for session resumptions.

Is MagicSock handshake protected from replay?

**Comment by Skidamek** [comment]
> TLS1.3 dropped support for session resumptions.

[It did not.](https://datatracker.ietf.org/doc/html/rfc8446#section-2.2)

**Comment by DXist** [comment]
Yes, indeed. It replaced stable session identifier like Session ID by an opaque ticket value which server could associate with the initial connection.

**Comment by matheus23** [comment]
There are two types of connections with iroh's relays:
- QAD (QUIC address discovery) connections we use as an encrypted alternative to STUN address discovery. QAD frames already return ASAP with the first packets from the relay IIRC. I thought that adding 0-RTT to these connections doesn't speed anything up? In any case - there's no application data transferred in these connections, so there's no danger from replay.
- HTTP/1.1 websocket connections from iroh endpoints to relays. I'm not 100% sure if 0-RTT is even possible in HTTP/1.1, but in any case this handshake can't really benefit from session resumption: It's either the server sending a challenge first, or it's the client sending a signed challenge from TLS extracted key material, which is only available after a full handshake.

**Comment by flub** [comment]
So this issue is about the HTTP/1.1 WebSocket connection. As you mention QAD is already as optimal as possible, adding 0-RTT to it does not achieve anything.

To my understanding session resumption shortens the TLS handshake. You indeed still have to perform the iroh-relay protocol handshake on the resumed session, but I don't think this resumed session stops you from extracting new key material from TLS.
